tests/melee_test.cpp

```diff
@@ -0,0 +1,244 @@
+#include "catch/catch.hpp"
+
+#include "npc.h"
+#include "monster.h"
+#include "monattack.h"
+#include "game.h"
+
+#include <string>
+#include <sstream>
+
+static float brute_probability( Creature& attacker, Creature &target, size_t iters )
+{
+    // Note: not using deal_melee_attack because it trains dodge, which causes problems here
+    size_t hits = 0;
+    for( size_t i = 0; i < iters; i++ ) {
+        int spread = attacker.hit_roll() - target.dodge_roll();
+        if( spread > 0 ) {
+            hits++;
+        }
+    }
+
+    return (float)hits / iters;
+}
+
+static float brute_special_probability( monster& attacker, Creature &target, size_t iters )
+{
+    size_t hits = 0;
+    for( size_t i = 0; i < iters; i++ ) {
+        if( !mattack::dodge_check( &attacker, &target ) ) {
+            hits++;
+        }
+    }
+
+    return (float)hits / iters;
+}
+
+static std::string full_attack_details( const player &dude )
+{
+    std::stringstream ss;
+    ss << "Details for " << dude.disp_name() << std::endl;
+    ss << "get_hit() == " << dude.get_hit() << std::endl;
+    ss << "get_hit_base() == " << dude.get_hit_base() << std::endl;
+    ss << "get_hit_weapon() == " << dude.get_hit_weapon( dude.weapon ) << std::endl;
+    return ss.str();
+}
+
+const int num_iters = 1000;
+
+TEST_CASE("Character attacking a zombie", "[melee]") {
+    monster zed( mtype_id( "mon_zombie" ) );
+    INFO( "Zombie has get_dodge() == " + std::to_string( zed.get_dodge() ) );
+
+    SECTION("8/8/8/8, no skills, unarmed") {
+        standard_npc dude( "TestCharacter", {}, 0, 8, 8, 8, 8 );
+        float prob = brute_probability( dude, zed, num_iters );
+        INFO( full_attack_details( dude ) );
+        THEN( "The chance to hit is near 60%" ) {
+            REQUIRE( prob > 0.5f );
+            REQUIRE( prob < 0.7f );
+        }
+    }
+
+    SECTION("8/8/8/8, 3 all skills, two-by-four") {
+        standard_npc dude( "TestCharacter", {}, 3, 8, 8, 8, 8 );
+        dude.weapon = item( "2x4" );
+        float prob = brute_probability( dude, zed, num_iters );
+        INFO( full_attack_details( dude ) );
+        THEN( "The chance to hit is near 70%" ) {
+            REQUIRE( prob > 0.6f );
+            REQUIRE( prob < 0.8f );
+        }
+    }
+
+    SECTION("10/10/10/10, 8 all skills, katana") {
```

Comment:
Sure, but we should still aim to define them all in one place. This is of benefit if we write further unit tests and/or expand the stats dumper. Being able to refer to S1, S2 .. S9 etc does help as well